lib/saluki-components/src/transforms/name_filter/mod.rs

```diff
@@ -0,0 +1,234 @@
+            metric_prefix_blacklist: self.metric_prefix_blacklist.clone(),
+            blocklist: Blocklist::new(&self.metric_blocklist, self.metric_blocklist_match_prefix),
+        }))
+    }
+}
+
+impl MemoryBounds for NameFilterConfiguration {
+    fn specify_bounds(&self, builder: &mut MemoryBoundsBuilder) {
+        // Capture the size of the heap allocation when the component is built.
+        builder.minimum().with_single_value::<NameFilter>();
+    }
+}
+
+#[derive(Debug)]
+struct Blocklist {
+    data: Vec<String>,
+    match_prefix: bool,
+}
+
+impl Blocklist {
+    fn new(data: &[String], match_prefix: bool) -> Self {
+        let mut data = data.to_owned();
+        data.sort();
+
+        // Only keep values with unique prefixes.
+        if match_prefix && !data.is_empty() {
+            let mut i = 0;
+            for j in 1..data.len() {
+                if !data[j].starts_with(&data[i]) {
+                    i += 1;
+                    data[i] = data[j].clone(); // Move item to next position
+                }
+            }
+            data.truncate(i + 1);
+        }
+
+        Blocklist { data, match_prefix }
+    }
+
+    fn contains(&self, name: &str) -> bool {
+        if self.data.is_empty() {
+            return false;
+        }
+
+        let i = self.data.binary_search(&name.to_string());
+
+        // Check for prefix match when match_prefix is true
+        if self.match_prefix {
+            let index = i.map_or_else(|idx| idx, |idx| idx);
+            if index > 0 && name.starts_with(&self.data[index - 1]) {
+                return true;
+            }
+        }
+
+        if let Ok(index) = i {
+            return name == self.data[index];
+        }
+
+        false
+    }
+}
+
+#[derive(Debug)]
+pub struct NameFilter {
+    metric_prefix: String,
+
+    metric_prefix_blacklist: Vec<String>,
+
+    blocklist: Blocklist,
+}
+
+impl NameFilter {
+    fn enrich_metric(&self, mut metric: Metric) -> Option<Metric> {
+        let metric_name = metric.context().name().deref();
+
+        if self.metric_prefix.is_empty() {
+            for s in &self.blocklist.data {
+                if s == metric_name {
+                    return None;
+                }
+            }
+        } else {
+            if self.is_excluded(metric_name) {
+                return None;
+            }
+
+            // Enrich metric with prefix and check if the blocklist contains the new name.
+            let new_metric_name = format!("{}{}", self.metric_prefix, metric_name);
+            if self.blocklist.contains(&new_metric_name) {
+                debug!("Metric {} excluded due to blocklist.", new_metric_name);
+                return None;
+            }
+
+            // Update metric with new name.
+            let new_context = metric.context().with_name(new_metric_name);
+            let existing_context = metric.context_mut();
+            *existing_context = new_context;
+        }
```

Comment:
Sorry, now that things are a little cleaner, I can see where the Datadog Agent code was going. We only want to prefix the metric if `self.metric_prefix` is actually set, _and_ if the original metric doesn't match any prefix configured in `self.metric_prefix_blacklist`.

As such, we want something like:

```rust
let new_metric_name = if self.is_excluded(...) {
    metric.context().name().clone()
} else {
    MetaString::from(format!(...))
}
```

That would end up generally being pretty efficient since the metric name should be inlined/interned so we're not allocating in the case where the metric has an excluded prefix.